File: frontend/ui/src/app/types/config-as-code.ts
export type ConfigAsCodeResourceKind =
  | 'DeploymentProcess'
  | 'Channel'
  | 'Lifecycle'
  | 'VariableSetDefinition'
  | 'StepTemplateReference'
  | 'Runbook';

export type ConfigAsCodeAuthorityValue = 'DATABASE_MANAGED' | 'GIT_MANAGED';

export interface ConfigAsCodeValidateRequest {
  documents: ConfigAsCodeValidateDocumentRequest[];
}

export interface ConfigAsCodeValidateDocumentRequest {
  content: string;
}

export interface ConfigAsCodeValidateResponse {
  valid: boolean;
  documents: ConfigAsCodeDocumentResult[];
  errors: ConfigAsCodeIssue[];
  warnings: ConfigAsCodeIssue[];
}

export interface ConfigAsCodeDocumentResult {
  kind: ConfigAsCodeResourceKind;
  apiVersion: string;
  metadataName?: string;
  metadataPath?: string;
  canonicalChecksum: string;
}

export interface ConfigAsCodeIssue {
  documentIndex: number;
  path: string;
  message: string;
}

export interface ConfigAsCodeAuthority {
  resourceKind: ConfigAsCodeResourceKind;
  resourceId: string;
  authority: ConfigAsCodeAuthorityValue;
  repositoryPath: string;
  sourceRevision: string;
  documentChecksum: string;
  updatedByUserId?: string;
  updatedAt: string;
}

export interface ConfigAsCodeAuthorityListResponse {
  authorities: ConfigAsCodeAuthority[];
}

export interface ConfigAsCodeAuthorityUpdateRequest {
  authority: ConfigAsCodeAuthorityValue;
  repositoryPath: string;
  sourceRevision: string;
  documentChecksum: string;
}

File: api/config_as_code.go
package api

type ConfigAsCodeValidateRequest struct {
	Documents []ConfigAsCodeValidateDocumentRequest `json:"documents"`
}

type ConfigAsCodeValidateDocumentRequest struct {
	Content string `json:"content"`
}

type ConfigAsCodeValidateResponse struct {
	Valid     bool                         `json:"valid"`
	Documents []ConfigAsCodeDocumentResult `json:"documents"`
	Errors    []ConfigAsCodeIssue          `json:"errors"`
	Warnings  []ConfigAsCodeIssue          `json:"warnings"`
}

type ConfigAsCodeDocumentResult struct {
	Kind              string `json:"kind"`
	APIVersion        string `json:"apiVersion"`
	MetadataName      string `json:"metadataName,omitempty"`
	MetadataPath      string `json:"metadataPath,omitempty"`
	CanonicalChecksum string `json:"canonicalChecksum"`
}

type ConfigAsCodeIssue struct {
	DocumentIndex int    `json:"documentIndex"`
	Path          string `json:"path"`
	Message       string `json:"message"`
}

type ConfigAsCodeAuthority struct {
	ResourceKind     string  `json:"resourceKind"`
	ResourceID       string  `json:"resourceId"`
	Authority        string  `json:"authority"`
	RepositoryPath   string  `json:"repositoryPath"`
	SourceRevision   string  `json:"sourceRevision"`
	DocumentChecksum string  `json:"documentChecksum"`
	UpdatedByUserID  *string `json:"updatedByUserId,omitempty"`
	UpdatedAt        string  `json:"updatedAt"`
}

type ConfigAsCodeAuthorityListResponse struct {
	Authorities []ConfigAsCodeAuthority `json:"authorities"`
}

type ConfigAsCodeAuthorityUpdateRequest struct {
	Authority        string `json:"authority"`
	RepositoryPath   string `json:"repositoryPath"`
	SourceRevision   string `json:"sourceRevision"`
	DocumentChecksum string `json:"documentChecksum"`
}
